modules/pulse/recorder.c

```diff
@@ -191,3 +195,75 @@ int pulse_recorder_alloc(struct ausrc_st **stp, const struct ausrc *as,
 
 	return err;
 }
+
+
+static void dev_list_cb(pa_context *c, const pa_source_info *l,
+						int eol, void *userdata)
+{
+	struct list *dev_list = userdata;
+	int err;
+
+	if (eol > 0) {
+		return;
+	}
+
+	/* In pulseaudio every sink automatically has a monitor source
+	This "output" device must be filtered out */
+	if (!strstr(l->name,"output")) {
+		err = mediadev_add(dev_list, l->name);
+		if (err) {
+			warning("pulse recorder: media device (%s) "
+					"can not be added\n",l->name);
+		}
+	}
+}
+
+
+static int set_available_devices(struct ausrc *as)
+{
+	int err;
+	pa_mainloop *pa_ml;
+	pa_mainloop_api *pa_mlapi;
+	pa_operation *pa_op;
+	pa_context *pa_ctx;
+
+	/* Create a mainloop API and connection to the default server */
+	pa_ml = pa_mainloop_new();
+	pa_mlapi = pa_mainloop_get_api(pa_ml);
+	pa_ctx = pa_context_new(pa_mlapi, "Baresip");
+
+	pa_context_connect(pa_ctx, NULL, 0, NULL);
+
+	while (pa_context_get_state(pa_ctx) != PA_CONTEXT_READY) {
+		err = pa_mainloop_iterate(pa_ml, 1, NULL);
+		if (err < 0) {
+			return err;
+		}
```

Comment:
when returning here, it will create a memory leak, the pa_op and pa_ctx will leak ..